File: 03 - source code/01 - imperative languages/Perl 5/random_streams_for_perf_stats.pl
# random_streams_for_perf_stats.pl
#
# 2025-06-01, 2025-07-21: better error handling when writing to files
# 2025-12-19: see below
# 2025-01-24: correcting the confusing variable naming (for experimental reasons back then)
#
# run on Ubuntu 24 LTS: $ perl random_streams_for_perf_stats.pl
#                       $ sudo perf stat -r 20 perl random_streams_for_perf_stats.pl
#
# $ perl --version
# This is perl 5, version 38, subversion 2 (v5.38.2) built for x86_64-linux-gnu-thread-multi
# $


use warnings;
use strict;
use IO::Handle;  # for flushing stdout


my constant $END = 62501;  # 62501 for exactly 1M binary digits
# my constant $END = 100;  # for testing

# my constant $M1   = 1_000_000;
# my constant $K250 =   250_000;

my constant $m = 65521;  # = 2^16 - 15
my constant $a = 17364;
my constant $c = 0;

my constant $file_bits_x   = 'random_bitstring.bin';
my constant $file_bits_hex = 'random_bitstring.byte';


my $random_start = int(rand($m - 1)) + 1;  # exclusive of m; 2025-12-19

my @x = ($random_start);  # also needed for the password later
my @bits_x;    # array of strings for the bit stream:  "0"'s + "1"'s
my @bits_hex;  # array of strings for the byte stream: [0..9][a..f]'s


print("\ngenerating a random bit stream...");
STDOUT->flush();

my $i = 1;
while ($i < $END) {
  my $y = ($a * $x[$i-1] + $c) % $m;
  push @x, $y;

  my $bits_x_str = sprintf('%016b', $y);  # Bin: 0001011111001100
  push @bits_x, $bits_x_str;

  my $bits_hex_str = sprintf('%04x', $y);  # Hex
  push @bits_hex, $bits_hex_str;

  $i++;
}

my $bits_x_str_total   = join('', @bits_x);
my $bits_hex_str_total = join('', @bits_hex);

# write bit stream to disk:
my $FH1_flag = 1;  # True = 1
unless(open(FH, '>', $file_bits_x)) {
  warn "\ncan't open file: $file_bits_x -- $!";
  $FH1_flag = 0;
}
unless(print FH $bits_x_str_total) {
  warn "could not write to file: $file_bits_x";
  $FH1_flag = 0;
}
unless(close(FH)) {
  warn "could not close file: $file_bits_x";
  $FH1_flag = 0;
}
STDOUT->flush();
if ($FH1_flag == 1) {
  print "\nBit stream has been written to disk under name:  $file_bits_x";
}

# write byte stream to disk:
my $FH2_flag = 1;  # True = 1
unless(open(FH, '>', $file_bits_hex)) {
  warn "\ncan't open file: $file_bits_hex -- $!";
  $FH2_flag = 0;
}
unless(print FH $bits_hex_str_total) {
  warn "could not write to file: $file_bits_hex";
  $FH2_flag = 0;
}
unless(close(FH)) {
  warn "could not close file: $file_bits_hex";
  $FH2_flag = 0;
}
STDOUT->flush();
if ($FH2_flag == 1) {
  print "\nByte stream has been written to disk under name: $file_bits_hex\n";
}

# end of random_streams_for_perf_stats.pl

File: 03 - source code/01 - imperative languages/Lua/random_streams_for_perf_stats.lua
-- random_streams_for_perf_stats.lua
--
-- 2025-07-13/14
--
-- run in Ubuntu 24 LTS: $ time lua random_streams_for_perf_stats.lua
--                       $ sudo perf stat -r 20 lua ./random_streams_for_perf_stats.lua
--                       
--                       $ time luajit random_streams_for_perf_stats.lua
--                       $ sudo perf stat -r 20 luajit ./random_streams_for_perf_stats.lua
--
-- $ lua -v
-- Lua 5.4.8  Copyright (C) 1994-2025 Lua.org, PUC-Rio
-- $ 
--
-- $ luajit -v
-- LuaJIT 2.1.1748459687 -- Copyright (C) 2005-2025 Mike Pall. https://luajit.org/
-- $


local math = require("math")
-- To optimize memory usage, prefer local variables over global ones.
-- Local variables are stored in the stack, which is faster and automatically cleaned up when no longer needed.
-- This practice minimizes the need for garbage collection.
-- https://dev.to/cristianalex_17/how-to-manage-memory-in-lua-in-2025-2a5b


------------------------------------------------------------------------
--
-- user defined functions

local STR_LENGTH_BIN = 16

local function Integer_to_bin_string(N)
    local bin_str = ""
    local j = STR_LENGTH_BIN
    local k = N
    while k > 0 and j >= 1 do
        if k % 2 == 0 then
            bin_str = '0'..bin_str
        else
            bin_str = '1'..bin_str
        end
        k = math.floor(k / 2)
        j = j - 1
    end
    bin_str = ('0'):rep(16-#bin_str)..bin_str
    return bin_str
end

-- end of user defined functions
--
------------------------------------------------------------------------


local END = 62500  -- 62500 for exactly 1M binary digits
-- local END = 10  -- for testing

local m = 65521  -- = 2^16 - 15
local a = 17364
local c = 0

local file_bits_x   = 'random_bitstring.bin'
local file_bits_hex = 'random_bitstring.byte'

local x = {}  -- also needed for the password later

math.randomseed(os.time())
x[1] = math.random(0, m)  -- exclusive of m

local bits_x   = {}  -- array of strings for the bit stream:  "0"'s + "1"'s
local bits_hex = {}  -- array of strings for the byte stream: [0..9][a..f]'s


print("\ngenerating a random bit stream...")

for i = 1, END do
  x[i+1] = (a * x[i] + c) % m
  -- print("\nx[i] = " .. x[i+1])  -- for testing

  local bits_x_str = Integer_to_bin_string(x[i+1])
  -- print("bits_x_str = " .. bits_x_str)  -- for testing
  bits_x[i] = bits_x_str

  local bits_hex_str = string.format("%04x", x[i+1])
  -- print("bits_hex_str = "..bits_hex_str)  -- for testing
  bits_hex[i] = bits_hex_str
end

local bits_x_str_total   = table.concat(bits_x, "")
local bits_hex_str_total = table.concat(bits_hex, "")
-- print("\nbits_x_str_total = "..bits_x_str_total)  -- for testing
-- print("\nbits_hex_str_total = "..bits_hex_str_total)  -- for testing

-- writing streams to files:
local file1 = io.open(file_bits_x, "w")
if file1 then
    file1:write(bits_x_str_total)
    file1:close()
    print("Bit stream has been written to disk under name:  "..file_bits_x)
else
    print("could not write to file: "..file_bits_x)
end

local file2 = io.open(file_bits_hex, "w")
if file2 then
    file2:write(bits_hex_str_total)
    file2:close()
    print("Byte stream has been written to disk under name: "..file_bits_hex)
else
    print("could not write to file: "..file_bits_hex)
end

-- end of random_streams_for_perf_stats.lua
